src/box/lua/net_replicaset.lua

```diff
@@ -0,0 +1,376 @@
+    local func = watcher.func
+    local instance_name = instance.instance_name
+    return function(key, value)
+        return func(key, value, instance_name)
+    end
+end
+
+-- Subscribe to `key` event on the leader of the replicaset.
+-- When the key value is updated the callback `func` will be called like that:
+-- func(key, value, instance_name).
+function replicaset_methods:watch_leader(key, func)
+    replicaset_check(self, 'watch_leader')
+    local watcher = {
+        replicaset = self,
+        key = key,
+        func = func,
+    }
+    self.leader_watchers[watcher] = watcher
+    local leader_instance = self.leader_instance
+    if leader_instance then
+        local wwfunc = wrap_watcher_function(watcher, leader_instance)
+        watcher.handle = leader_instance.conn:watch(watcher.key, wwfunc)
+    end
+    return setmetatable(watcher, watcher_mt)
+end
+
+-- Unregister all leader watchers from old leader (leader count becomes not 1).
+local function replicaset_unbind_watchers(replicaset)
+    for _, watcher in pairs(replicaset.leader_watchers) do
+        watcher.handle:unregister()
+        watcher.handle = nil
+    end
+end
+
+-- Register all leader watchers to the new leader (leader count becomes 1).
+local function replicaset_rebind_watchers(replicaset)
+    local leader_instance = replicaset.leader_instance
+    for _, watcher in pairs(replicaset.leader_watchers) do
+        local wwfunc = wrap_watcher_function(watcher, leader_instance)
+        watcher.handle = leader_instance.conn:watch(watcher.key, wwfunc)
+    end
+end
+
+-- Close all backend connections.
+function replicaset_methods:close()
+    replicaset_check(self, 'close')
+    for _, instance in pairs(self.instances) do
+        instance.conn:close()
+    end
+end
+
+-- Callback that called if the status of an instance is (possibly) changed.
+-- Called from watcher ('ro' or 'rw') or from on_disconnect ('unknown').
+local function instance_status_change(replicaset, instance, status)
+    if status == instance.status then
+        return
+    end
+    replicaset.wait_status_cond:broadcast()
+    local status_count = replicaset.status_count
+    status_count[instance.status] = status_count[instance.status] - 1
+    instance.status = status
+    status_count[instance.status] = status_count[instance.status] + 1
+
+    if status_count.rw ~= 1 and replicaset.leader_instance then
+        replicaset.leader_instance = nil
+        replicaset_unbind_watchers(replicaset)
+    elseif status_count.rw == 1 and not replicaset.leader_instance then
+        if status == 'rw' then
+            replicaset.leader_instance = instance
+        else
+            for _, search_instance in pairs(replicaset.instances) do
+                if search_instance.status == 'rw' then
+                    replicaset.leader_instance = search_instance
+                end
+            end
+        end
+        replicaset_rebind_watchers(replicaset)
```

Comment:
A good thing about watchers is that it's guaranteed that a watcher callback is never executed in parallel with itself for the same key. This simplifies programming callbacks because one doesn't have to bother about synchronization.

Unregistering a net.box watcher doesn't stop the currently running callback. As a result, during leader switching it may occur that the same watcher callback is running in two fibers - for the old leader and for the new one.

Reply:
Here we can use a condvar (for example)